CGPs/cgp-0031.md

```diff
@@ -16,27 +16,112 @@ This proposal sets the initial parameters for and effectively activates Granda M
+  * This is proposed to be 2 weeks (a value of `1209600` seconds).
+  * The current governance parameters involve a 1 day approval stage and 1 week referendum stage. Two weeks is chosen to allow ample time to create a veto proposal and gather community feedback, while also balancing needs from exchangers. This delay also serves as friction, imposing an opportunity cost by locking capital to prevent exchangers from using Granda Mento for a means of arbitrage or trivial intents.
 3. **`FixidityLib.Fraction public spread;`** - The percent fee imposed upon an exchange execution.
+  * The proposed spread is 0% (a value of `0`).
   * Given the opportunity cost imposed upon exchangers when using Granda Mento, there is already a disincentive to use Granda Mento. Granda Mento requires sufficient friction while also being accessible compared to other stablecoin issuers at scale, such as USDC and USDT. USDT charges a [0.1% fee](https://tether.to/fees/); however they have much lower limits ($100k) and close to immediate redemption. USDC is [largely free](https://www.circle.com/en/pricing), with some fees associated with card and bank wire transactions, also with immediate redemption. Given this landscape of competing fees, the proposed spread is 0%.
-4. **`mapping(string => ExchangeLimits) stableTokenExchangeLimits`** - Set by Governance via `setStableTokenExchangeLimits`. A mapping of the stable token registry ID string to a struct containing two `uint256`s, `minExchangeAmount` and `maxExchangeAmount`.
+4. **`FixidityLib.Fraction public maxApprovalExchangeRateChange;`** - The maximum change of the CELO price quoted in the stable token between the time an exchange proposal is created and the time the exchange proposal is approved by the approver.
+  * This is proposed to be 100% (a value of `1000000000000000000000000`).
+  * The purpose of this parameter is twofold:
+    * To protect the exchanger in a situation where the sell asset's price is locked in when it is very low, and the price has since risen substantially prior to approval.
+    * To protect the reserve in a situation where the sell asset's price is locked in when it is very high, and the price has since fallen substantially prior to approval.
+      * A similar situation could also occur in a long-range attack, where a proposal is intentionally created when the price of the sell asset is high, locking in the high price. Then, at a point when the price of the sell asset is lower, the proposer can bring the proposal to the community, hoping the community does not notice the high price.
+  * Prior CELO price action was analyzed on Coinbase Pro to come to the 100% figure. Two examples of dramatic price action stand out:
+    * A prominent example of CELO going from a high to low price:
+      * May 16, 2021 (+0 days): high of $6.98 (-0%)
+      * May 19, 2021 (+3 days): low of $3.70 (-88%)
```

Comment:
what number this -88% is calculated on? I'd presume is $6.98, but going from $6.98 to $3.70 is not a 88% decline. I think I must be missing something

Reply:
Lol oops I did it relative to the lower value because I thought for some reason the solidity code did that-- this means that a value of 100% would protect from price increases but not price decreases. I'll come up with a new value

Reply:
There probably isn't enough liquidity on Moola (or of wCELO on Eth platforms) but in theory someone could:
i) Buy CELO, plus simultaneously borrow an equal amount of CELO (collateralised by stables) = a net neutral CELO position.
ii) Apply for Granda Mento at their CELO purchase price.
iii) If CELO price decreases (i.e. cUSD price increases, denominated in CELO), proceed with Granda Mento at the locked in price, then sell cUSD back to CELO at market price (liquidity permitting), then close out the borrow position.

I think this is somewhat theoretical because of liquidity constraints, but wanted to put this out there.

Intuitively, I would think putting a CELO price change limit (between proposal and execution) of maybe 5% seems about right. It will result in some proposals being cancelled, but maybe that's ok.

Probably also an idea for us to keep an eye on liquidity for CELO borrowing on Moola and also wCELO borrowing (I'm not sure if that's even possible on Aave or Compound).

Reply:
@tkporter , ok those graphs are illuminating.

It seems 50% seems reasonable for a starting point. It's probably a good thing that the minimum exchange value is pretty high as that makes it hard to arbitrage by borrowing in parallel on Moola.

Should we be making this value dynamic - or do we just start like this and then do another governance proposal to lower/adjust it in future if need be? I think that's fine, just want us to be aware that this 50% is being hard coded in for now and I'm guessing will need to be changed some day.

What does Zviad think about the value?

Reply:
I think that the implicit assumption that past historical data informs future price action is a dangerous one. Today is an example of "dramatic price action" that could get a proposal thrown out by the system even if a proposal was made in earnest at a much lower CELO price. After the fact, the proposer could just resubmit at a much higher price to their benefit (and the Reserve's loss). I think this type of range might be better evaluated on a case by case basis by the multisig in conjunction with the nature of the proposal rather than having a hard coded implementation. A wild deviation in price might be a red flag but shouldn't be the sole reason for disqualifying an exchange, as market pricing may be *very* unpredictable.

Reply:
@Pinotio at the moment it would require a future governance proposal to adjust in the future if needed.

@albertclabs yeah, in retrospect I think this parameter probably introduces unnecessary constraints / pain when configuring Granda Mento. I've been thinking of this parameter as a very high upper bound on price changes as a very general safety check that the reserve can't take a *crazy* loss, but ultimately I think it's on the approver signers to ensure that Granda Mento won't be abused. So if a proposal not violating the `maxApprovalExchangeRateChange` doesn't mean it's in the clear to be approved -- instead I'd expect the approver signers to take a look at the market context & decide if the exchange is not totally fair.